pytorch_lightning/callbacks/gpu_stats_monitor.py

```diff
@@ -0,0 +1,164 @@
+    ):
+        super().__init__()
+
+        if shutil.which("nvidia-smi") is None:
+            raise MisconfigurationException(
+                'Cannot use GPUStatsMonitor callback because NVIDIA driver is not installed.'
+            )
+
+        self._log_stats = Namespace(
+            memory_utilization=memory_utilization,
+            gpu_utilization=gpu_utilization,
+            intra_step_time=intra_step_time,
+            inter_step_time=inter_step_time,
+            fan_speed=fan_speed,
+            temperature=temperature
+        )
+
+    def on_train_batch_start(self, trainer, pl_module, batch, batch_idx, dataloader_idx):
+        if self._log_stats.gpu_utilization:
+            self._log_gpu(trainer)
+        if self._log_stats.memory_utilization:
+            self._log_memory(trainer)
+
+        if self._log_stats.inter_step_time and self.snap_inter_step_time:
+            # First log at beginning of second step
+            trainer.logger.log_metrics(
+                {'batch_time/inter_step (ms)': (time.time() - self.snap_inter_step_time) * 1000},
+                step=trainer.global_step
+            )
+
+        if self._log_stats.intra_step_time:
+            self.snap_intra_step_time = time.time()
+
+    def on_train_batch_end(self, trainer, pl_module, batch, batch_idx, dataloader_idx):
+        if self._log_stats.gpu_utilization:
+            self._log_gpu(trainer)
+
+        if self._log_stats.memory_utilization:
+            self._log_memory(trainer)
+
+        if self._log_stats.fan_speed:
+            trainer.logger.log_metrics(self._get_gpu_stat("fan.speed", "%"), step=trainer.global_step)
+
+        if self._log_stats.temperature:
+            trainer.logger.log_metrics(self._get_gpu_stat("temperature.gpu", "degrees C"), step=trainer.global_step)
+            trainer.logger.log_metrics(self._get_gpu_stat("temperature.memory", "degrees C"), step=trainer.global_step)
+
+        if self._log_stats.inter_step_time:
+            self.snap_inter_step_time = time.time()
+
+        if self._log_stats.intra_step_time and self.snap_intra_step_time:
+            trainer.logger.log_metrics(
+                {'batch_time/intra_step (ms)': (time.time() - self.snap_intra_step_time) * 1000},
+                step=trainer.global_step
+            )
+
+    def on_train_start(self, trainer, pl_module):
+        if not trainer.logger:
+            raise MisconfigurationException(
+                'Cannot use GPUStatsMonitor callback with Trainer that has no logger.'
+            )
+
+        if not trainer.on_gpu:
+            rank_zero_warn(
+                'You are using GPUStatsMonitor but are not running on GPU.'
+                ' Logged utilization will be independent from your model.', RuntimeWarning
+            )
+
+    def on_train_epoch_start(self, trainer, pl_module):
+        self.snap_intra_step_time = None
+        self.snap_inter_step_time = None
+
+    @staticmethod
+    def _get_gpu_stat(pitem: str, unit: str):
```

Comment:
could we get all inf in a single cal? because this kind of system call can delay training...

Reply:
yeah, will try to update this.

Reply:
this is not critical, you can just add todo note and resolve it in other PR :]